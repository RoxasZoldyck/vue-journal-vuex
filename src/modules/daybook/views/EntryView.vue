<!-- Los views siempre son componentes, pero la diferencia entre los demás view es que estos componentes se usan para navegar entre diferentes rutas, algo así como un contenedor de otro componentes -->

<template>
    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{dayMonthYear.day}}</span>
                <span class="mx-1 fs-3">{{dayMonthYear.month}}</span>
                <span class="mx-2 fs-4 fw-light">{{dayMonthYear.yearDay}}</span>
            </div>

            <!--El botón de borrar sólo se mostrará si hay una entrada, validamos con entry.id-->
            <div>
                <input type="file"
                @change="onSelectedImage"
                ref="imageSelector"
                v-show="false"
                accept="image/png, image/jpeg"> <!--Este input no aparecerá, pero con la ref lo utilizaremos cuando se presiona el botón de subir foto-->

                <button 
                v-if="entry.id"
                class="btn btn-danger mx-2"
                @click="onDeleteEntry">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <button class="btn btn-primary"
                @click="onSelectImage"> <!--Con esta función buscaremos la referencia del input-->
                    Subir foto
                    <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>

        <!-- Esto es exclusivo de vue 3, llamar otro root element -->
        <hr>
        
        <div class="d-flex flex-column px-3 h-75">
            <textarea 
            v-model="entry.text"
            placeholder="¿Qué sucedió hoy?"></textarea>
        </div>

        <Fab 
        icon="fa-save"
        @on:click="saveEntry"
        /> <!--El icon se está enviando como prop al Fab, indicando así que cambie el icono del botón Fab-->

        <img 
        v-if="entry.picture && !localImage"
        :src="entry.picture"
        alt="entry-picture"
        class="img-thumbnail">
        
        <img 
        v-if="localImage"
        :src="localImage"
        alt="entry-picture"
        class="img-thumbnail">

    </template>
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapActions,mapGetters } from 'vuex'; //computed
import Swal from 'sweetalert2'

import getDayMonthYear from '../helpers/getDayMonthYear'
import uploadImage from '../helpers/uploadImage'

export default {
    name:'EntryView',
    props:{
        id:{
            type:String,
            required:true
        }
    },
    components:{
        Fab: defineAsyncComponent(() => import('@/modules/daybook/components/Fab.vue'))
    },
    data(){
        return{
            entry:null,
            localImage:null,
            file:null
            // entry:{
            //     text:''
            // }
        }
    },
    //Mi respuesta:
    computed:{
        ...mapGetters('journalModule',['getEntryById']), //Llamamos al getter
        dayMonthYear(){ //Esta función da el formato correcto a las fechas obtenidas por this.entry.date
            const {day, month, yearDay} = getDayMonthYear(this.entry.date)
            return {day,month,yearDay}
        }
        // day(){
        //     console.log(this.entry.date)
        //     const {day} = getDayMonthYear(this.entry.date)
        //     return day
        // },
        // month(){
        //     const {month} = getDayMonthYear(this.entry.date)
        //     return month
        // },
        // yearDay(){
        //     const {yearDay} = getDayMonthYear(this.entry.date)
        //     return yearDay
        // },

    },
    methods:{
        loadEntry(){
            //Esto corrige el bug de que cada que se cambie de imagen se mantenga la imagen
            this.localImage = null
            this.file = null

            let entry //Declaramos entry

            if(this.id === 'new'){ //Validamos si el id 'new' que es cuando se presiona el botón de '+'
                entry = {
                    text: '',
                    date: new Date().getTime()
                }
            }else{ //Sino obtenemos la entrada seleccionada
                entry = this.getEntryById(this.id)
                if(!entry) return this.$router.push({name:'no-entry'}) //Si no hay información de la entrada seleccionada, redirigimos a fuera de la entrada

                // return this.getEntryById(this.id)
            }
            this.entry = entry //Al data.entry le damos la información del entry


            
        },
        ...mapActions('journalModule', ['updateEntry','createEntry','deleteEntry']),
        async saveEntry(){
            //Creamos una alerta de espera en lo que se guarda una entrada
            Swal.fire({
                title:'Espere por favor',
                allowOutsideClick: false
            })

            Swal.showLoading()

            //Llamamos a la función que sube la imagen en el servidor
            const picture = await uploadImage(this.file)

            //Una vez subida la imagen al servidor, retorna el url en el que se está guardando, es tan simple como asignarlo al this.entry.picture
            this.entry.picture = picture

            if (this.entry.id){ //Si this.entry.id tiene un valor, significa que quiere actualizarse    
                // console.log(this.entry)
                //Action del Journal Module
                await this.updateEntry(this.entry)
            }else{
                //Mi respuesta
                //Sino lo tiene, significa que quiere crear una nueva entrada
                // console.log('Post de una nueva entrada')

                //await action
                const id = await this.createEntry(this.entry) //Aquí se está guardando lo del return del actions

                //redirectTo => entry, param:id
                this.$router.push({name:'entry', params:{id}})
                
            }

            this.file = null //Para vaciar de nuevo la data y que no se sobreponga la imagen cuando se cambie de entrada
            Swal.fire('Guardado', 'Entrada registrada con éxito', 'success')
        },
        async onDeleteEntry(){ //Creamos la función para borrar entrada
            const {isConfirmed} = await Swal.fire({
                title:'¿Está seguro?',
                text:'Una vez borrado, no se puede recuperar',
                showDenyButton:true,
                confirmButtonText:'Sí, estoy seguro'
            })

            if (isConfirmed){ //isConfirmed es un bool que devuelve await Swal.fire cuando se presiona que sí se desea borrar
                //Mostramos alerta de espera
                Swal.fire({
                    title:'Espere por favor',
                    allowOutsideClick: false
                })
                Swal.showLoading()

                //Borramos la entrada enviando el entry.id a las acción deleteEntry
                await this.deleteEntry(this.entry.id)

                //redirectTo => no-entry
                this.$router.push({name:'no-entry'})
                
                //Mostramos que se completó la eliminación
                Swal.fire('Eliminado','','success')
            }


        },
        onSelectedImage(event){
            const file = event.target.files[0] //Por defecto event.target.files devuelve un arreglo, en este caso sólo accederemos al primero
            
            //Si no tenemos un archivo, entonces simplemente convertimos en null nuestras variables de data relacionadas con la imagen
            if(!file){
                this.localImage = null
                this.file = null
                return
            }

            //Usamos esto para guardar el file que se sube en una variable local de la data
            this.file = file

            //Lo siguiente carga el archivo en la variable localImage
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL(file)
        },
        onSelectImage(){
            // console.log(this.$refs)
            this.$refs.imageSelector.click() //Esto es de js, es como si hiciéramos un: document.querySelector('input').click(), entonces hace que el subir foto funcione como el input
            
        }
    },
    created(){
        // console.log(this.$route.params.id)
        // console.log(this.id)
        this.loadEntry() //Llamamos al método que a su vez llama al getEntryById
    },

    watch:{
        id(){ //Observará el id, y lo que esté dentro sólo se ejecutará cuando este id haya cambiado
            this.loadEntry()
        }
    }

    //Curso:
    // computed:{
    //     ...mapGetters('journalModule',['getEntryById']),
    // },
    // methods:{
    //     loadEntry(){
    //         const entry = this.getEntryById(this.id)
    //         console.log(entry)
    //     }
    // },    
    // created(){
    //     this.loadEntry()
    // }
}
</script>

<style lang="scss" scoped>
//Estilos para que el textarea no tenga bordes
    textarea{
        font-size: 20px;
        border: none;
        height: 100%;

        &:focus{
            outline: none;
        }
    }
    img{
        width: 200px;
        position: fixed;
        bottom: 150px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
    }
</style>