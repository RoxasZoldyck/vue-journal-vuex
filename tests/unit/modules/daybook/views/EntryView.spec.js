import { shallowMount } from "@vue/test-utils"
import { createStore } from "vuex"
import journalModule from "@/modules/daybook/store/journal"
import { journalState } from "../../../mock-data/test-journal-state"
import EntryView from "@/modules/daybook/views/EntryView.vue"

import Swal from 'sweetalert2'


const createVuexStore = (initialState) =>
    createStore({
        modules: {
            journalModule: {
                ...journalModule, //Desestructuramos todo lo que journalModule contenga
                state: { ...initialState } //Sobreescribimos el state con lo que haya en initialState
            }
        }
    })

//Creamos un mock de sweetaler2
jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {

    const store = createVuexStore(journalState) //Creamos nuestro store virtual
    store.dispatch = jest.fn() //Aquí estamos reemplazando la función del dispatch, y asignándole una jest.fn(), así nos aseguramos de que no se borre de Firebase y todo se quede en un entorno virtual

    const mockRouter = {
        push: jest.fn()
    } //Creamos el router virtual, este básicamente se encarga de que se cumplan $router

    let wrapper

    //Hacemos esto para que el wrapper se reinicie cada que se ejecute una prueba
    beforeEach(() => {
        jest.clearAllMocks() //Limpiamos todos los mocks
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-NxHcYiqqeYjJDIJT7G8'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })
    })

    test('debe de sacar al usuario porque el id no existe', () => {
        const wrapper = shallowMount(EntryView, {
            props: {
                id: 'Este ID no existe en el store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        })

        expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' }) //Aquí estamos enviando un id que no existe y validamos que nos mande a no-entry

    })

    test('debe de mostrar la entrada correctamente', () => {
        expect(wrapper.html()).toMatchSnapshot() //Validamos la snapshot haya sido cargada (Que se haya cargado el wrapper)
        expect(mockRouter.push).not.toHaveBeenCalled() //Sí es así, entonces también validamos que el mockRouter no haya sido llamado
    })

    test('debe de borrar la entrada y salir', (done) => {
        Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true })) //Aquí estamos simulando que se ejecute Swal y regrese una vez con isConfirmed:true

        wrapper.find('.btn-danger').trigger('click') //Buscamos el botón de borrar entrada y mandamos un click virtual

        expect(Swal.fire).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado, no se puede recuperar',
            showDenyButton: true,
            confirmButtonText: 'Sí, estoy seguro'
        }) //Nos aseguramos que el Swal.fire virtual haya sido llamado con el objeto deseado

        setTimeout(() => { //Se usa el Timeout por un error que hay
            expect(store.dispatch).toHaveBeenCalledWith("journalModule/deleteEntry", "-NxHcYiqqeYjJDIJT7G8") //Validamos que se haya llamado con la función deleteEntry y se pase como argumento el id
            expect(mockRouter.push).toHaveBeenCalled()//Validamos de que el push al router se haya llamado (Una vez que se elimina una entrada, te envía a no-entry)
            done()
        }, 1)
    })
})