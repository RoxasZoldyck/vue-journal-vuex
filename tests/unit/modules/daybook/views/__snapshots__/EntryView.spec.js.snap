// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">7</span><span class="mx-1 fs-3">Mayo</span><span class="mx-2 fs-4 fw-light">2024, Martes</span></div>
  <!--El botón de borrar sólo se mostrará si hay una entrada, validamos con entry.id-->
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;">
    <!--Este input no aparecerá, pero con la ref lo utilizaremos cuando se presiona el botón de subir foto--><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary">
      <!--Con esta función buscaremos la referencia del input--> Subir foto <i class="fa fa-upload"></i>
    </button>
  </div>
</div>
<!-- Esto es exclusivo de vue 3, llamar otro root element -->
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?"></textarea></div>
<fab-stub icon="fa-save"></fab-stub>
<!--El icon se está enviando como prop al Fab, indicando así que cambie el icono del botón Fab-->
<img src="https://res.cloudinary.com/dqkcsldfl/image/upload/v1715083085/t2inytxfmokfuojkl5pu.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
`;
